File: hermes/enterprise/token_crypto.py
"""AES-256-GCM token decryption — native port of Divo's ``token.crypto.ts``.

Divo stores Zoho/Google OAuth tokens encrypted at rest in Postgres. The
transformed Hermes runtime reads those rows directly and decrypts them in
process, so this must match Divo's scheme byte-for-byte.

Reference: ``advance-backend/src/infrastructure/shared/token.crypto.ts``

Key derivation:
  - ``base64:<...>`` prefix → base64-decode the remainder (must be 32 bytes).
  - otherwise → SHA-256 of the raw string → 32-byte key.

Cipher text format: ``v1:<iv_b64>:<tag_b64>:<ciphertext_b64>``
  - IV: 12 bytes (96-bit GCM nonce)
  - tag: 16 bytes (GCM auth tag), stored separately from ciphertext
  - the Python AESGCM API expects ``ciphertext || tag``, so we append the tag

The encryption key env var is ``ZOHO_TOKEN_ENCRYPTION_KEY`` (same key Divo uses
for both Zoho and Google in production).

Decrypt-only by design: this phase treats the credential store as read-only, so
no ``encrypt_token`` is provided. If/when Hermes owns the write path (token
refresh persisted back to Postgres), add the matching encrypt here.
"""

from __future__ import annotations

import base64
import hashlib
import os

DEFAULT_KEY_ENV = "ZOHO_TOKEN_ENCRYPTION_KEY"
_CIPHER_PREFIX = "v1"
_BASE64_KEY_PREFIX = "base64:"


class TokenCryptoError(RuntimeError):
    """Raised when a token cannot be decrypted (bad key, format, or data)."""


def _to_key_bytes(raw: str) -> bytes:
    raw = raw.strip()
    if not raw:
        raise TokenCryptoError("Token encryption key is empty or not configured")
    if raw.startswith(_BASE64_KEY_PREFIX):
        key = base64.b64decode(raw[len(_BASE64_KEY_PREFIX):])
        if len(key) != 32:
            raise TokenCryptoError("base64 encryption key must resolve to 32 bytes")
        return key
    # SHA-256 of the raw string → deterministic 32-byte key
    return hashlib.sha256(raw.encode("utf-8")).digest()


def resolve_key(encryption_key: str | None = None) -> bytes:
    """Resolve the 32-byte AES key from an explicit value or the env var."""
    raw = encryption_key if encryption_key is not None else os.getenv(DEFAULT_KEY_ENV, "")
    return _to_key_bytes(raw or "")


def decrypt_token(cipher_text: str, encryption_key: str | None = None) -> str:
    """Decrypt a ``v1:<iv>:<tag>:<data>`` token. Raises ``TokenCryptoError``."""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    raw = (cipher_text or "").strip()
    parts = raw.split(":")
    if len(parts) != 4 or not parts[0].startswith("v"):
        raise TokenCryptoError(
            "Invalid encrypted token format (expected v1:<iv>:<tag>:<data>)"
        )

    key = resolve_key(encryption_key)
    try:
        iv = base64.b64decode(parts[1])
        tag = base64.b64decode(parts[2])
        data = base64.b64decode(parts[3])
    except (ValueError, base64.binascii.Error) as exc:  # type: ignore[attr-defined]
        raise TokenCryptoError("Encrypted token contained invalid base64") from exc

    try:
        # Python's AESGCM expects ciphertext concatenated with the 16-byte tag.
        plaintext = AESGCM(key).decrypt(iv, data + tag, None)
    except Exception as exc:  # noqa: BLE001 — normalize to one error type
        raise TokenCryptoError(
            "Token decryption failed — key mismatch or data corrupted"
        ) from exc

    return plaintext.decode("utf-8")


def try_decrypt_token(
    cipher_text: str | None, encryption_key: str | None = None
) -> str | None:
    """Decrypt, returning ``None`` for empty input or on failure (never raises)."""
    if not cipher_text:
        return None
    try:
        return decrypt_token(cipher_text, encryption_key)
    except TokenCryptoError:
        return None

File: advance-backend/src/infrastructure/shared/token.crypto.ts
/**
 * token.crypto — AES-256-GCM token encryption/decryption.
 *
 * Shared by backend-owned integration credential storage.
 *
 * Key derivation:
 *   - If the raw key starts with "base64:" → parse the remainder as base64 (must be 32 bytes).
 *   - Otherwise → SHA-256 hash the string → 32-byte key.
 *
 * Cipher text format: `v1:<iv_b64>:<tag_b64>:<ciphertext_b64>`
 *
 * Legacy v1 integration rows use ZOHO_TOKEN_ENCRYPTION_KEY. New provider rows
 * may use INTEGRATION_TOKEN_ENCRYPTION_KEY, selected by their persisted
 * tokenCipherVersion so key rotation does not make existing ciphertext unreadable.
 */

import { createCipheriv, createDecipheriv, createHash, randomBytes } from 'node:crypto';

// ─── Constants ────────────────────────────────────────────────────────────────

const CIPHER_ALGO    = 'aes-256-gcm';
const IV_LENGTH      = 12; // 96-bit nonce
const CIPHER_VERSION = 1;

// ─── Error ────────────────────────────────────────────────────────────────────

export class TokenCryptoError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TokenCryptoError';
  }
}

// ─── Key derivation ────────────────────────────────────────────────────────────

function toKeyBuffer(raw: string): Buffer {
  if (raw.startsWith('base64:')) {
    const buf = Buffer.from(raw.slice('base64:'.length), 'base64');
    if (buf.length !== 32) {
      throw new TokenCryptoError('base64 encryption key must resolve to 32 bytes');
    }
    return buf;
  }
  // SHA-256 of the raw string → deterministic 32-byte key
  return createHash('sha256').update(raw).digest();
}

function readKey(encryptionKey: string): Buffer {
  const raw = encryptionKey.trim();
  if (!raw) {
    throw new TokenCryptoError('Token encryption key is empty or not configured');
  }
  const key = toKeyBuffer(raw);
  if (key.length !== 32) {
    throw new TokenCryptoError('Token encryption key must resolve to 32 bytes');
  }
  return key;
}

// ─── Encrypt ──────────────────────────────────────────────────────────────────

export interface EncryptedToken {
  /** The versioned cipher text: `v1:<iv_b64>:<tag_b64>:<data_b64>` */
  cipherText: string;
  version: number;
}

export function encryptToken(plainText: string, encryptionKey: string): EncryptedToken {
  const value = plainText.trim();
  if (!value) {
    throw new TokenCryptoError('Cannot encrypt an empty token');
  }

  const key = readKey(encryptionKey);
  const iv  = randomBytes(IV_LENGTH);

  const cipher    = createCipheriv(CIPHER_ALGO, key, iv);
  const encrypted = Buffer.concat([cipher.update(value, 'utf8'), cipher.final()]);
  const tag       = cipher.getAuthTag();

  return {
    version:    CIPHER_VERSION,
    cipherText: `v${CIPHER_VERSION}:${iv.toString('base64')}:${tag.toString('base64')}:${encrypted.toString('base64')}`,
  };
}

// ─── Decrypt ──────────────────────────────────────────────────────────────────

export function decryptToken(cipherText: string, encryptionKey: string): string {
  const raw   = cipherText.trim();
  const parts = raw.split(':');

  if (parts.length !== 4 || !(parts[0] ?? '').startsWith('v')) {
    throw new TokenCryptoError('Invalid encrypted token format (expected v1:<iv>:<tag>:<data>)');
  }

  const key     = readKey(encryptionKey);
  const iv      = Buffer.from(parts[1] ?? '', 'base64');
  const authTag = Buffer.from(parts[2] ?? '', 'base64');
  const data    = Buffer.from(parts[3] ?? '', 'base64');

  const decipher = createDecipheriv(CIPHER_ALGO, key, iv);
  decipher.setAuthTag(authTag);

  try {
    const decrypted = Buffer.concat([decipher.update(data), decipher.final()]);
    return decrypted.toString('utf8');
  } catch {
    throw new TokenCryptoError('Token decryption failed — key mismatch or data corrupted');
  }
}
